### PR TITLE
Added back the ZbootStatus handlers to the main eventloop

### PR DESCRIPTION
**Pull request by srinibas-zededa**
2. Added back the ZbootStatus handlers to the main eventloop.
This will ensure, we do receive the events and also post them to
the controller.

Signed-off-by: Srinibas Maharana <srinibas@zededa.com>

### CODE REVIEW

**Comment by kalyan-nidumolu** [comment]
> 2\. Added back the ZbootStatus handlers to the main eventloop.

Hi Srinibas,

Can you explain what is the functional impact of this fix? What is the bug you are trying to fix?

Thanks,
Kalyan

File: pkg/pillar/cmd/zedagent/zedagent.go [hunk 1]
```diff
@@ -849,27 +864,49 @@ func Run() {
...
+	if ctxPtr.devInfoTriggerFail {
+		return
+	}
 	log.Info("Triggered PublishDeviceInfo")
 	select {
 	case ctxPtr.TriggerDeviceInfo <- struct{}{}:
 		// Do nothing more
 	default:
-		log.Info("Failed to send on PublishDeviceInfo")
+		log.Errorf("Failed to send on PublishDeviceInfo")
```
Comment (kalyan-nidumolu) [inline pkg/pillar/cmd/zedagent/zedagent.go]: Where do we see the error as to why the publish failed? Would be good to print it here..

**Comment by kalyan-nidumolu** [comment]
"Added a retry logic, when the triggerPublichDevInfo fails" --> Is this not already done based on the timer? Why do we need to explicitly retrigger the publish again??

"Added back the ZbootStatus handlers to the main eventloop.
This will ensure, we do receive the events and also post them to
the controller."

--> What was missing before this?

File: pkg/pillar/cmd/zedagent/zedagent.go [hunk 2]
```diff
@@ -849,27 +864,49 @@ func Run() {
...
+		triggerPublishDevInfo(ctxPtr)
 	}
 }
 
 func triggerPublishDevInfo(ctxPtr *zedagentContext) {
 
+	// last publish event has to be complete,
+	// for handling current publish event trigger,
+	// wait until then
+	if ctxPtr.devInfoTriggerFail {
```
Comment (kalyan-nidumolu) [inline pkg/pillar/cmd/zedagent/zedagent.go]: Hmm.. Should you be checking for devInfoTaskReady here?

  if ctxPtr.devInfoTaskReady {
      return
 }

File: pkg/pillar/cmd/zedagent/zedagent.go [hunk 3]
```diff
@@ -880,6 +917,11 @@ func deviceInfoTask(ctxPtr *zedagentContext, triggerDeviceInfo <-chan struct{})
 		case <-stillRunning.C:
 		}
 		agentlog.StillRunning(agentName + "devinfo")
+
+		// mark self as ready to receive publish trigger again
+		if !ctxPtr.devInfoTaskReady {
```
Comment (kalyan-nidumolu) [inline pkg/pillar/cmd/zedagent/zedagent.go]: The if check may be redundant? Always mark yourself ready ??? You are setting it to true in any case?

File: pkg/pillar/cmd/zedagent/zedagent.go [hunk 4]
```diff
@@ -98,6 +98,8 @@ type zedagentContext struct {
 	rebootCmd                 bool
 	rebootCmdDeferred         bool
 	deviceReboot              bool
+	devInfoTriggerFail        bool
+	devInfoTaskReady          bool
```
Comment (kalyan-nidumolu) [inline pkg/pillar/cmd/zedagent/zedagent.go]: Can you add comments on what each of these used for?

devInfoTaskReady --> This indicates if a publish is currently in progress - right?
   - So, this must be set just before Publish is called - And reset after Publish is Done??

devInfoTriggerFail
   - This is really an indication if the last publish failed - right? And if it failed, you want to retry the publish after some trigger.. right?
   - So, should we name it as devInfoTaskSuccess? Success if True / Failed if False?

File: pkg/pillar/cmd/zedagent/zedagent.go (hunk 1 above)
Comment (srinibas-zededa) [reply pkg/pillar/cmd/zedagent/zedagent.go]: The device is generating multiple events which end up in issuing triggers for publishDevInfo. This is good to keep a tab on the failures while issuing triggers for publistDevInfo.

File: pkg/pillar/cmd/zedagent/zedagent.go (hunk 2 above)
Comment (srinibas-zededa) [reply pkg/pillar/cmd/zedagent/zedagent.go]: This is avoid, queing more triggers to channel. We have a retry mechanism, when the trigger processing is over.

File: pkg/pillar/cmd/zedagent/zedagent.go (hunk 4 above)
Comment (srinibas-zededa) [reply pkg/pillar/cmd/zedagent/zedagent.go]: sure. will add appropriate comments.

**Comment by srinibas-zededa** [comment]
> I do not understand why these changes are needed at all. What is the failure?
> 
> The triggerDeviceInfo intentionally has a buffer of 1 meaning that if the task has not yet started or is busy formatting and sending a device info message, there can still be one trigger queued. In such a case the task will send a device info message (again or when it starts).
> 
> Do we forget to pick up zbootStatus early in the life of zedagent?

@nordmark , we are seeing failures in scheduling triggers in to channel. While multiple events are getting generated during device bootup or upgrade, some of the events trigger the publishDevInfo, and they are failing.
Also, the zbootStatus handler code was removed in the main eventloop , during refactoring for nodeagent. 
This code changes, addresses the above two issues.

**Comment by kalyan-nidumolu** [comment]
> > I do not understand why these changes are needed at all. What is the failure?
> > The triggerDeviceInfo intentionally has a buffer of 1 meaning that if the task has not yet started or is busy formatting and sending a device info message, there can still be one trigger queued. In such a case the task will send a device info message (again or when it starts).
> > Do we forget to pick up zbootStatus early in the life of zedagent?
> 
> @nordmark , we are seeing failures in scheduling triggers in to channel. While multiple events are getting generated during device bootup or upgrade, some of the events trigger the publishDevInfo, and they are failing.

[Kalyan] Not sure I understand what you are saying. What are you trying to avoid here? 
   "seeing failures in scheduling triggers in to channel." --> What does this mean? Why are these failures happening and are they expected? And if a failure happens, doesn't the next trigger make it right? And then there is the timer triggering the publish anyways..

> Also, the zbootStatus handler code was removed in the main eventloop , during refactoring for nodeagent.
> This code changes, addresses the above two issues.

[Kalyan] What was the functionality bug that is caused by this?

**Comment by eriknordmark** [comment]
> @nordmark , we are seeing failures in scheduling triggers in to channel. While multiple events are getting generated during device bootup or upgrade, some of the events trigger the publishDevInfo, and they are failing.

If the code calls trigger 17 times and 16 of them don't send on the channel that is perfectly OK (in fact it is a key part of the design) as long as the final state is published.
Do you have a case where the final state is not published?

> Also, the zbootStatus handler code was removed in the main eventloop , during refactoring for nodeagent.

OK, I don't have an issue with the second part.

**Comment by srinibas-zededa** [comment]
> > @nordmark , we are seeing failures in scheduling triggers in to channel. While multiple events are getting generated during device bootup or upgrade, some of the events trigger the publishDevInfo, and they are failing.
> 
> If the code calls trigger 17 times and 16 of them don't send on the channel that is perfectly OK (in fact it is a key part of the design) as long as the final state is published.
> Do you have a case where the final state is not published?
When there is no more events getting triggered (for some time),  and somehow, we miss the final trigger for the device state change, the final state will not get published. Anyways, I have reverted the  retry logic.  
> 
> > Also, the zbootStatus handler code was removed in the main eventloop , during refactoring for nodeagent.
> 
> OK, I don't have an issue with the second part.

**Pull request merged**